File: engine/models/trade.py
from dataclasses import dataclass, field
from datetime import datetime
from typing import Optional
from engine.models.signal import SignalType


@dataclass
class Trade:
    id: str
    strategy_id: str
    symbol: str
    side: SignalType
    qty: float
    entry_price: float
    exit_price: Optional[float] = None
    opened_at: datetime = field(default_factory=datetime.utcnow)
    closed_at: Optional[datetime] = None
    pnl: Optional[float] = None
    fees: float = 0.0

    @property
    def is_open(self) -> bool:
        return self.exit_price is None

    def close(self, exit_price: float, fees: float = 0.0) -> None:
        self.exit_price = exit_price
        self.closed_at = datetime.utcnow()
        self.fees += fees
        multiplier = 1 if self.side == SignalType.BUY else -1
        self.pnl = multiplier * (exit_price - self.entry_price) * self.qty - self.fees

File: dashboard/src/types/trade.ts
export interface Trade {
  id: string;
  strategy_id: string;
  symbol: string;
  side: "buy" | "sell" | "close";
  qty: number;
  entry_price: number;
  exit_price?: number;
  opened_at: string;
  closed_at?: string;
  pnl?: number;
  fees: number;
}

export interface FillEvent {
  type: "fill";
  strategy_id: string;
  order_id: string;
  signal: "buy" | "sell";
  price: number;
  qty: number;
}

export type WsEvent = FillEvent | { type: string; [key: string]: unknown };
